ironfish/src/syncer.ts

```diff
@@ -123,16 +162,90 @@ export class Syncer {
       .getConnectedPeers()
       .filter((peer) => peer.features?.syncing && peer.work && peer.work > head.work)
 
-    // Get a random peer with higher work. We do this to encourage
-    // peer diversity so the highest work peer isn't overwhelmed
-    // as well as helping to make sure we don't get stuck with unstable peers
-    if (peers.length > 0) {
-      const peer = ArrayUtils.sampleOrThrow(peers)
+    if (peers.length === 0) {
+      return
+    }
+
+    // If there is only one valid peer to sync from, there is no point in
+    // measuring the connection so begin syncing immediately
+    if (peers.length === 1) {
+      this.startSync(peers[0])
+      return
+    }
+
+    const syncCandidates = ArrayUtils.shuffle(peers).slice(0, 8)
+
+    // If we have been syncing from a peer, we want to include this peer in the
+    // measurement. This will allow us to maintain a connection to a strong peer
+    // if we have found one.
+    if (
+      currentPeerIdentity &&
+      !syncCandidates.find((p) => p.getIdentity() === currentPeerIdentity)
+    ) {
+      const currentPeer = this.peerNetwork.peerManager.getPeer(currentPeerIdentity)
+      if (currentPeer && currentPeer.state.type === 'CONNECTED') {
+        syncCandidates.push(currentPeer)
+      }
+    }
+
+    const peerRtt = new Map<Identity, number>()
+
+    // Measure how long it takes to fetch the genesis block header from each
+    // peer as an estimate of connection quality
+    for (const peer of syncCandidates) {
+      if (this.state === 'stopped' || this.state === 'stopping') {
+        return
+      }
+
+      const peerIdentity = peer.getIdentity()
+      if (peerIdentity == null) {
+        continue
+      }
+
+      const start = BenchUtils.start()
+      try {
+        const response = await this.peerNetwork.getBlockHeaders(peer, 1, 1)
```

Comment:
I think this could probably be explored separately, but I wonder if we could speed up measurement by using `Promise.any` to measure two (or more) peers at once, resolving to the one that completes first. It would use a bit more bandwidth and as such make the test a bit less reliable, but I expect it would reduce the impact of peers that are unusually slow (since their response would get dropped)

Reply:
I did consider that, but I felt it was more correct to do them sequentially. In situations where my connection is sufficiently capable, then having simultaneous results would probably not affect the results much. However, if we assume my connection is the bottle-neck, then having multiple requests out simultaneously would definitely taint the results.

It's worth discussing further for sure, as I'm making assumptions here that I haven't _fully_ thought out / tested. During testing though, most of the time spent was the 2 seconds between stopping syncing and the next event loop execution, barring the very occasional response timeout. Rahul and I tested on a c6i.large in Australia, and the results were basically the same. Admittedly, these aren't extremely thorough tests, and the network seems to have a high percentage of powerful, capable peers to sync from right now.

Reply:
Sounds good! If this hasn't been an issue in testing, let's keep it as-is and make changes if necessary.